index.js.flow

```diff
@@ -453,53 +451,66 @@ export type Projection = {
   ...
 };
 
-type AggregationUpdate = (
+type AggregationUpdate = $ReadOnlyArray<
   | {|
-      ['$addFields' | '$set' | '$replaceWith']: { [string]: Expression, ... },
+      +['$addFields' | '$set' | '$replaceWith']: { [string]: Expression, ... },
     |}
   | {|
-      $project:
-        | { [string]: 0 | false, _id?: Expression | 0 | false | 1 | true, ... }
-        | { [string]: Expression | 1 | true, _id?: Expression | 0 | false | 1 | true, ... },
+      +$project:
+        | { +[string]: 0 | false, +_id?: Expression | 0 | false | 1 | true, ... }
+        | { +[string]: Expression | 1 | true, +_id?: Expression | 0 | false | 1 | true, ... },
     |}
-  | {| $unset: string | string[] |}
-  | {| $replaceRoot: { newRoot: { [string]: Expression, ... } } |}
-)[];
-
-type AtomicUpdate = {
-  [
-    | '$addToSet'
-    | '$bit'
-    | '$currentDate'
-    | '$inc'
-    | '$max'
-    | '$min'
-    | '$mul'
-    | '$pop'
-    | '$pull'
-    | '$pullAll'
-    | '$push'
-    | '$rename'
-    | '$set'
-    | '$setOnInsert'
-    | '$unset']: mixed,
-  ...,
-};
+  | {| +$unset: string | string[] |}
+  | {| +$replaceRoot: { +newRoot: { +[string]: Expression, ... } } |}
+>;
+
+export type Update = {|
+  $addToSet?: { [string]: mixed, ... },
+  $bit?: { ['and' | 'or' | 'xor']: number },
+  $currentDate?: { [string]: true | { $type: 'timestamp' | 'date' } },
+  $inc?: { [string]: number, ... },
+  $max?: { [string]: mixed, ... },
+  $min?: { [string]: mixed, ... },
+  $pop?: { [string]: -1 | 1, ...  },
+  $pull?: { [string]: mixed, ... },
+  $pullAll?: { [string]: mixed, ... },
+  $push?: { [string]: mixed, ... },
+  $rename?: { [string]: mixed, ... },
+  $set?: { [string]: mixed, ... },
+  // TODO: forbid this option if upsert is not true.
+  $setOnInsert?: { [string]: mixed, ... },
+  $unset?: { [string]: mixed, ... },
+|};
+
+type ReadOnlyUpdate = $ReadOnly<{
```

Comment:
"ReadOnly" + "Update" could be a bit confusing out of context. Not an issue if the type isn't exposed to the user.

Reply:
Yeah I hear that - I'll add a clarifying comment.

Reply:
cd5181ecf93c002b89a3feb8e1c5466ca449d6ec